source/common/upstream/upstream_connection_pool_impl.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <unordered_map>
+
+#include "envoy/api/api.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/filter.h"
+#include "envoy/upstream/upstream.h"
+
+#include "common/common/linked_object.h"
+#include "common/common/thread.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class UpstreamConnectionPoolImpl : public UpstreamConnectionPool {
```

Comment:
+1

Reply:
In general can you add a big comment somewhere about the threading model? It would be nice to have an overview of what the connection pool thread does, etc.

Edit from reading: I'm a little confused why we need a thread for the shared connection pool. Can't all of the events just happen via inline callbacks based on adopt/offer APIs? Or is the issue that you need a dispatcher to deal with cleanup properly based on the way all of the other code works?

Reply:
We need a dispatcher to detect upstream disconnects while the connection is idle, otherwise we may try to reuse a connection that has been closed for minutes.  We could consider keeping the upstream connection registered for disconnects in the original worker thread instead of using a global dispatcher and have workers go to the right peer worker to retrieve the pooled connection, but that approach seems more complicated.

Reply:
OK that makes sense. I think some high level comments and more comments in general would help a lot.